async-server.c

```diff
@@ -0,0 +1,727 @@
+
+    pthread_mutex_lock(&context.tasks_mutex);
+    pthread_cond_wait(&context.tasks_cond, &context.tasks_mutex);
+    pthread_mutex_unlock(&context.tasks_mutex);
+
+    memcpy(task->password, context.password, sizeof(context.password));
+
+    pthread_mutex_lock(&context.set_mutex);
+    for (int i = 0; i < context.set.size; ++i)
+    {
+        pthread_t receiver_thread = context.set.data[i].receiver_id;
+        pthread_cancel(receiver_thread);
+        pthread_join(receiver_thread, NULL);
+
+        pthread_t sender_thread = context.set.data[i].sender_id;
+        pthread_cancel(sender_thread);
+        pthread_join(sender_thread, NULL);
+
+        close_client(context.set.data[i].socket_fd);
+    }
+    pthread_mutex_unlock(&context.set_mutex);
+
+    pthread_cancel(server_thread);
+    pthread_join(server_thread, NULL);
+
+    queue_destroy(&context.queue);
+    set_destroy(&context.set);
+    close(server_socket);
+
+    return context.found;
+}
+
+/***************
+ * CLIENT CODE *
+ ***************/
+
+struct cl_context_t
+{
+    struct queue_t queue;
+    struct queue_t queue_done;
+
+    volatile int tasks_done;
+    volatile int tasks_running;
+    pthread_mutex_t tasks_mutex;
+
+    volatile bool server_done;
+
+    int server_sfd;
+    struct config_t *config;
+};
+
+static int
+read_message(int socket_fd, void *data)
+{
+    int status;
+
+    int length;
+    status = recvall(socket_fd, &length, sizeof(length), 0);
+    if (status == -1) return -1;
+
+    status = recvall(socket_fd, data, length, 0);
+    if (status == -1) return -1;
+
+    return 0;
+}
+
+static void *
+cl_worker(void *arg)
+{
+    struct cl_context_t *context = (struct cl_context_t *) arg;
+    struct config_t *config = context->config;
+
+    struct st_context_t st_context;
+    st_context.hash = config->hash;
+    st_context.cd.initialized = 0;
+
+    while (true)
+    {
+        struct task_t task;
+        queue_pop(&context->queue, &task);
+
+        bool found = process_task(&task, config,
+                                  &st_context, st_password_handler);
+        task.correct = found;
+
+        pthread_mutex_lock(&context->tasks_mutex);
+        --context->tasks_running;
+        ++context->tasks_done;
+        pthread_mutex_unlock(&context->tasks_mutex);
+
+        queue_push(&context->queue_done, &task);
+    }
+    return NULL;
+}
+
+static void *
+cl_task_receiver(void *arg)
+{
+    struct cl_context_t *context = arg;
+    int server_sfd = context->server_sfd;
+
+    int status;
+    struct task_t task;
+    while (true)
+    {
+        enum command_t tag;
+        status = recvall(server_sfd, &tag, sizeof(tag), 0);
+        if (status == -1) break;
+
+        switch (tag)
+        {
+        case CMD_TASK:
+            status = read_message(server_sfd, &task);
+            if (status == -1) goto exit_label;
+
+            pthread_mutex_lock(&context->tasks_mutex);
+            queue_push(&context->queue, &task);
```

Comment:
Looks suspicious. I would expect queue_push to be out of mutex lock. Lock is required only for tasks_running increment.